QMapboxGLSettings functions for changing default options and getting default styles.

kiibimees opened the pull request:
Since MapLibre is now supporting default settings for different providers. It must be possible to switch between different default settings from Qt plugin.

Previously, default map styles where defined in the Qt plugin source code. MapLibre has multiple provider support and has default style list for most common ones already defined. Adding possibility to get this information and use in Qt plugin. Hence, removing hard-coded default styles from Qt plugin source code.

ntadej commented on .gitmodules:
```diff
@@ -18,7 +18,7 @@
 [submodule "vendor/boost"]
 	path = vendor/boost
-	url = https://github.com/mapbox/mapbox-gl-native-boost.git
+	url = https://github.com/kiibimees/mapbox-gl-native-boost.git
```
We should probably ask for a proper MapLibre fork.

kiibimees replied:
Yes, I agree. It's discussed here https://github.com/maplibre/maplibre-gl-native/issues/124 but it doesn't progress, for some reason :)

kiibimees replied:
I would be so glad if any of the maintainers resolve https://github.com/maplibre/maplibre-gl-native/issues/124 and we could point this submodule to a more correct repository.

ntadej commented on CMakeLists.txt:
```diff
@@ -3,14 +3,15 @@ cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
 option(MBGL_WITH_CORE_ONLY "Build only the core bits, no platform code" OFF)
 option(MBGL_WITH_COVERAGE "Enable code coverage collection" OFF)
 option(MBGL_WITH_QT "Build Mapbox GL Qt bindings" OFF)
+option(MBGL_QT_STATIC "Build Mapbox GL Qt bindings staticly" OFF)
```
Why not make a more general flag like `MBGL_OBJECT_LIBRARY` or something? Looks a bit strange in a group of all `MBGL_WITH_*`.

kiibimees replied:
I moved this option from `platform/qt/qt.cmake`. I agree that naming can be better but the general idea is to always build one bundled library. When building a shared library, all mbgl libs can be STATIC. When building a staic library (iOS for example), all mbgl libraries have to be OBJECT because this is the simplest way to generate one final, single file, bundled lib.

kiibimees replied:
So, I would leave it like this because the intension is not to buil OBJECT library. The solution is to build temporary OBJECT libraries when the final Qt library is requested to be static. Hence, the naming is quite correct.

ntadej replied:
I do not like polluting global CMake configuration with Qt stuff as there are probably other clients of this, but let's keep it as this for now as the build system anyways needs a bit of cleanup in the future.

ntadej commented on CMakeLists.txt:
```diff
@@ -3,14 +3,15 @@ cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
 
 if (MBGL_WITH_QT)
-    set(CMAKE_OSX_DEPLOYMENT_TARGET 10.13)
+    set(CMAKE_OSX_DEPLOYMENT_TARGET 13.0)
```
This sounds a bit too new. Why do you need to change it?

kiibimees replied:
Because 13.0 is now Qt minimum. It can be taht it's min for Qt 6 only. I guess a if/else could be places here?

ntadej replied:
I guess this will not work for macOS as it's still on 12.1?

kiibimees replied:
Hmm... no idea. My main target is Qt 6. I guess the easiest would be to just to check Qt version here.

kiibimees replied:
Reverted this change and moved created override set(...) into qt.cmake where Qt version is already known.

ntadej replied:
This still looks like 13.0 to me.

ntadej replied:
Not sure what happened here, but we can fix it in a follow-up.

kiibimees replied:
It is already fixed by the [latter commit](https://github.com/maplibre/maplibre-gl-native/pull/197/commits/7391fc2b3ad4a16d873f4589d3e1d3170179f782) in this MR.

ntadej commented on platform/qt/qt.cmake:
```diff
@@ -7,21 +7,27 @@ set(MBGL_QT_VERSION_COMPATIBILITY 2.0.0)
 
-option(MBGL_QT_STATIC "Build Mapbox GL Qt bindings staticly" OFF)
 option(MBGL_QT_LIBRARY_ONLY "Build only libraries" OFF)
+option(MBGL_QT_INSIDE_QT_PLUGIN "Build qmapboxgl as OBJECT library, so it can be bundled into separate single plugin lib." OFF)
```
I guess we may still want to support Qt 5 for some time where object libraries would not work.

kiibimees replied:
Why object libraries would not work for Qt5?

ntadej replied:
I thought object libraries only work with CMake. If this also works with qmake, then it's fine.

ntadej replied:
MapLibre is mostly a drop-in replacement for MapBox (see https://github.com/ntadej/qt-geoservices-maplibre). If we use the old qmake-based build infrastruture for Qt 5 plugins, we still need to be able to link to some sort of static maplibre-gl-native library or libraries.

kiibimees replied:
A, ok, I see. But it should work because OBJECT library is only internal to the CMake build tree. Final lib will always be either a shared or static `qmapboxgl`.

kiibimees commented:
> Also I suspect the location you now set is a bit too late in the build (CMake is a bit sensitive). Maybe we could just not set it and load `QtAutoDetect` if inside Qt build tree?

It's not too late in the build. I just tested it and it works. It also works when not settings this at all (when building Qt plugin inside QtLocation build tree). Why was it set there in the first place? Perhaps we could indeed remove this totally?

ntadej commented:
> > Also I suspect the location you now set is a bit too late in the build (CMake is a bit sensitive). Maybe we could just not set it and load `QtAutoDetect` if inside Qt build tree?
> 
> It's not too late in the build. I just tested it and it works. It also works when not settings this at all (when building Qt plugin inside QtLocation build tree). Why was it set there in the first place? Perhaps we could indeed remove this totally?

I think Qt5 CMake files do not set the minimum platform versions so by default the latest is used. Probably it's hard to fix properly at this point and we'll just need to set the defaults reasonably.

kiibimees commented:
> I think Qt5 CMake files do not set the minimum platform versions so by default the latest is used. Probably it's hard to fix properly at this point and we'll just need to set the defaults reasonably.

Yeah, and if you look at my last commit, I just solved the issue so that if it's not set yet, it will be set. Othewise don't change.

rinigus commented:
I have tested this on Linux. As it looks to be super-seeding #192, I was mainly interested whether it would work for making plugins based on it. 

Plugin tests were done by compiling, installing the lib, compiling Mapbox GL QML plugin, and running Pure Maps using it.

`-DMBGL_WITH_QT_HEADLESS=OFF -DMBGL_QT_LIBRARY_ONLY=ON -DMBGL_QT_STATIC=ON`: works as expected

`-DMBGL_WITH_QT_HEADLESS=OFF -DMBGL_QT_LIBRARY_ONLY=ON -DMBGL_QT_STATIC=OFF`:  works as expected

Those above were tested for Sailfish OS builds using OBS and on phone - all worked as expected. Note that Sailfish is using Qt 5.6 (yes, you read it correctly) and it is not clear when we will update. So, would be good to keep that compatibility in MapLibre as well.

Next, static library was built at OBS for range of Sailfish OS versions for testing with older compilers and cmake. Here, there is a setback with `cmake 3.11` failing at configuration stage with
```
[   39s] + CFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[   39s] + export CFLAGS
[   39s] + CXXFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[   39s] + export CXXFLAGS
[   39s] + FFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables -I/usr/lib/gfortran/modules'
[   39s] + export FFLAGS
[   39s] + cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DCMAKE_SKIP_RPATH:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DMBGL_WITH_QT=ON -DMBGL_WITH_WERROR=OFF -DCMAKE_INSTALL_PREFIX:PATH=/usr -DMBGL_WITH_QT_HEADLESS=OFF -DMBGL_QT_LIBRARY_ONLY=ON -DMBGL_QT_STATIC=ON .
[   39s] -- The CXX compiler identification is GNU 8.3.0
[   39s] -- The C compiler identification is GNU 8.3.0
[   39s] -- Check for working CXX compiler: /usr/bin/c++
[   39s] -- Check for working CXX compiler: /usr/bin/c++ -- works
[   39s] -- Detecting CXX compiler ABI info
[   40s] -- Detecting CXX compiler ABI info - done
[   40s] -- Detecting CXX compile features
[   40s] -- Detecting CXX compile features - done
[   40s] -- Check for working C compiler: /usr/bin/cc
[   40s] -- Check for working C compiler: /usr/bin/cc -- works
[   40s] -- Detecting C compiler ABI info
[   40s] -- Detecting C compiler ABI info - done
[   40s] -- Detecting C compile features
[   40s] -- Detecting C compile features - done
[   40s] -- Configuring GL-Native with OpenGL renderer backend
[   40s] CMake Error at vendor/csscolorparser.cmake:16 (target_link_libraries):
[   40s]   Object library target "mbgl-vendor-csscolorparser" may not link to
[   40s]   anything.
[   40s] Call Stack (most recent call first):
[   40s]   CMakeLists.txt:981 (include)
[   40s] 
[   40s] 
[   40s] -- Configuring incomplete, errors occurred!
[   40s] See also "/home/abuild/rpmbuild/BUILD/qmaplibregl-2.0.0+master.20220109141436.c3d798c/maplibre-gl-native/CMakeFiles/CMakeOutput.log".

```

That failure was not there for shared library config, in agreement with your IF statement https://github.com/kiibimees/maplibre-gl-native/blob/qt6_plugin/vendor/csscolorparser.cmake#L5.

For shared libs, I have error (again for `cmake 3.11`, newer ones as `3.19` are doing fine):
```
[  110s] + CFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[  110s] + export CFLAGS
[  110s] + CXXFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[  110s] + export CXXFLAGS
[  110s] + FFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables -I/usr/lib/gfortran/modules'
[  110s] + export FFLAGS
[  110s] + LD_AS_NEEDED=1
[  110s] + export LD_AS_NEEDED
[  110s] + CFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[  110s] + export CFLAGS
[  110s] + CXXFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables'
[  110s] + export CXXFLAGS
[  110s] + FFLAGS='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -m32 -msse -msse2 -march=i686 -mfpmath=sse -mtune=generic -fno-omit-frame-pointer -fasynchronous-unwind-tables -I/usr/lib/gfortran/modules'
[  110s] + export FFLAGS
[  110s] + cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DCMAKE_SKIP_RPATH:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DMBGL_WITH_QT=ON -DMBGL_WITH_WERROR=OFF -DCMAKE_INSTALL_PREFIX:PATH=/usr -DMBGL_WITH_QT_HEADLESS=OFF -DMBGL_QT_LIBRARY_ONLY=ON .
[  111s] -- The CXX compiler identification is GNU 8.3.0
[  111s] -- The C compiler identification is GNU 8.3.0
[  111s] -- Check for working CXX compiler: /usr/bin/c++
[  111s] -- Check for working CXX compiler: /usr/bin/c++ -- works
[  111s] -- Detecting CXX compiler ABI info
[  111s] -- Detecting CXX compiler ABI info - done
[  111s] -- Detecting CXX compile features
[  111s] -- Detecting CXX compile features - done
[  111s] -- Check for working C compiler: /usr/bin/cc
[  111s] -- Check for working C compiler: /usr/bin/cc -- works
[  111s] -- Detecting C compiler ABI info
[  111s] -- Detecting C compiler ABI info - done
[  111s] -- Detecting C compile features
[  112s] -- Detecting C compile features - done
[  112s] -- Configuring GL-Native with OpenGL renderer backend
[  112s] -- Configuring GL-Native with Qt bindings
[  112s] -- Build type: Release
[  112s] -- Version 2.0.0
[  112s] -- Found the following ICU libraries:
[  112s] --   uc (required)
[  112s] -- Found ICU: /usr/include (found version "66.1") 
[  112s] CMake Error at platform/qt/qt.cmake:267 (install):
[  112s]   install TARGETS given no LIBRARY DESTINATION for shared library target
[  112s]   "qmapboxgl".
[  112s] Call Stack (most recent call first):
[  112s]   CMakeLists.txt:1051 (include)
```

That destination error was fixed in https://github.com/rinigus/maplibre-gl-native/commit/e8b14b6d16f586b1fea0be5965786f904815b76c and, with that, parity with Mapbox GL Native was restored for the shared lib. 

In addition, tried to compile for more common case:

**default options:** compilation **failed** as in 
```
[ 97%] Linking CXX executable mbgl-test-runner
/usr/bin/cmake -E cmake_link_script CMakeFiles/mbgl-test-runner.dir/link.txt --verbose=1
/usr/bin/c++ -O3 -DNDEBUG CMakeFiles/mbgl-test-runner.dir/platform/qt/test/main.cpp.o -o mbgl-test-runner  -Wl,--whole-archive test/libmbgl-test.a -Wl,--no-whole-archive libmbgl-core.a /usr/lib64/libQt5OpenGL.so.5.15.2 /usr/lib64/libQt5Widgets.so.5.15.2 /usr/lib64/libQt5Gui.so.5.15.2 libmbgl-vendor-csscolorparser.a libmbgl-vendor-parsedate.a -lz /usr/lib64/libQt5Network.so.5.15.2 /usr/lib64/libQt5Sql.so.5.15.2 /usr/lib64/libQt5Core.so.5.15.2 /usr/lib64/libicuuc.so -ldl libmbgl-vendor-nunicode.a test/libmbgl-vendor-googletest.a 
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: test/libmbgl-vendor-googletest.a(gmock-all.cc.o): undefined reference to symbol 'pthread_key_delete@@GLIBC_2.2.5'
/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: /lib64/libpthread.so.0: error adding symbols: DSO missing from command line
```
Adding `-pthread` for linking fixed it.

kiibimees commented:
Welll, so this is why (for older CMake versions) this explicit DESTINATION set was needed:
```
RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
FRAMEWORK DESTINATION "${CMAKE_INSTALL_LIBDIR}"
LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
```
I will put them back.

About `Object library target "mbgl-vendor-csscolorparser" may not link to anything.`, it seems that older CMake doesn't see behind `$<BUILD_INTERFACE:...>` generator expression. I wonder what happend when you remove generator expression and just link to object libs like this:

kiibimees commented:
> I wonder what happens when you remove generator expression and just link to object libs like this:

Actually, it will not work. It's different in CMake 3.11 and older: https://cmake.org/cmake/help/v3.11/manual/cmake-buildsystem.7.html#object-libraries

kiibimees commented:
@rinigus , I think I have found a solution for older CMake versions. It works with newer versions but I haven't actually testes on 3.11. Could you please test this: https://github.com/maplibre/maplibre-gl-native/pull/197/commits/23c0e2f3053aa08378a25b2f4742ea07aafcf8b4

rinigus commented:
@kiibimees: I will start testing right away. Will report back

rinigus commented:
Shared lib is building nicely now. For static, though, I am still getting

```
[   42s] -- The CXX compiler identification is GNU 8.3.0
[   42s] -- The C compiler identification is GNU 8.3.0
[   42s] -- Check for working CXX compiler: /usr/bin/c++
[   42s] -- Check for working CXX compiler: /usr/bin/c++ -- works
[   42s] -- Detecting CXX compiler ABI info
[   42s] -- Detecting CXX compiler ABI info - done
[   43s] -- Detecting CXX compile features
[   43s] -- Detecting CXX compile features - done
[   43s] -- Check for working C compiler: /usr/bin/cc
[   43s] -- Check for working C compiler: /usr/bin/cc -- works
[   43s] -- Detecting C compiler ABI info
[   43s] -- Detecting C compiler ABI info - done
[   43s] -- Detecting C compile features
[   43s] -- Detecting C compile features - done
[   43s] -- Configuring GL-Native with OpenGL renderer backend
[   43s] CMake Error at vendor/csscolorparser.cmake:16 (target_link_libraries):
[   43s]   Object library target "mbgl-vendor-csscolorparser" may not link to
[   43s]   anything.
[   43s] Call Stack (most recent call first):
[   43s]   CMakeLists.txt:981 (include)
```

So, that part of the commit didn't help, unfortunately.

In a grand scheme of things I think it is OK. I can use shared lib for older SFOS releases as well.

kiibimees commented:
Ou, yes, indeed. I forgot to add these OBJECT libs as TARGET_OBJECTS sources because for some strange reason it worked without it. I hope this fixes the problem. @rinigus, Try this one: https://github.com/maplibre/maplibre-gl-native/pull/197/commits/537c50c73f11d1e05f78638f76e65831f5bed839

rinigus commented:
@kiibimees: sorry, I am getting exactly the same error for static builds on cmake 3.11:

```
[   43s] -- Configuring GL-Native with OpenGL renderer backend
[   43s] CMake Error at vendor/csscolorparser.cmake:16 (target_link_libraries):
[   43s]   Object library target "mbgl-vendor-csscolorparser" may not link to
[   43s]   anything.
[   43s] Call Stack (most recent call first):
[   43s]   CMakeLists.txt:981 (include)
```

Notice that the error is at `vendor/csscolorparser.cmake:16` - not sure it is correct location, but that keeps repeating.

rinigus commented:
Just to recap - we have all working fine with the shared lib version, even cmake 3.11. For static builds, newer cmake is fine, cmake 3.11 complains.

Unless something obvious, we could proceed with such combination. I am not sure there are many users with old cmake that must use static builds.

kiibimees commented:
Ou, yes. I'm sorry. I understood the error message a bit wrong. The problem is NOT that it's not linked to anything. The problem is that it IS linked but should not be. I set up CMake 3.11 myself and tried to solve the issue but then understood that it would need a lot of changes in different CMake files. I have to agree that we could leave it like this.

I reverted my humble try about the linkage workaround and squashed a few of the last commits and force-pushed.

rinigus commented:
Retested compilation on different SFOS versions after your forced push and it works for me as expected. Hopefully, someone with the "keys" can review and merge it.

kiibimees commented:
@rinigus  or @ntadej , can't you add your review? It seems that I can at least review some other pull requests.

ntadej commented:
We need someone from the maintainers to push the pipeline. Unfortunately they have been a bit inactive lately.

rinigus commented:
@kiibimees: there are some review comments above by @ntadej that seem to be unresolved. Maybe they should be marked if resolved?

@ntadej - do you know who to ping? maybe we should just do it over here?

kiibimees commented:
I closed all conversations but one: The Boost fix and submodule fork needs also maintainers.

petr-pokorny-1 commented:
> We need someone from the maintainers to push the pipeline. Unfortunately they have been a bit inactive lately.

I just pushed the pipeline.

petr-pokorny-1 commented:
> We need someone from the maintainers to push the pipeline. Unfortunately, they have been a bit inactive lately.

@ntadej I am wondering if you have permission to kick off the pipeline. You are not a first-time contributor so you should be able to do it. At least that is what I see in the project settings. 

![Screenshot 2022-01-19 at 22 37 30](https://user-images.githubusercontent.com/16223511/150218244-966c84cb-941b-4267-9a67-8b5f7525b2f7.png)

Could you check it? I can investigate more and fix the settings. It is silly that you cannot launch pipelines!

ntadej commented:
> > We need someone from the maintainers to push the pipeline. Unfortunately, they have been a bit inactive lately.
> 
> @ntadej I am wondering if you have permission to kick off the pipeline. You are not a first-time contributor so you should be able to do it. At least that is what I see in the project settings.
> 
> ![Screenshot 2022-01-19 at 22 37 30](https://user-images.githubusercontent.com/16223511/150218244-966c84cb-941b-4267-9a67-8b5f7525b2f7.png)
> 
> Could you check it? I can investigate more and fix the settings. It is silly that you cannot launch pipelines!

It's not my MR and only maintainers can approve new contributor (which I am not): "First-time contributors need a maintainer to approve running workflows."

petr-pokorny-1 commented:
> It's not my MR and only maintainers can approve new contributor (which I am not): "First-time contributors need a maintainer to approve running workflows."

I see. I don't seem to have permission to add you to MapLibre organization. I will ask someone on Slack to do it.

rinigus commented:
So, as far as I can see, we wait now till #124 is resolved. That's the only outstanding review change. Which, again, requires someone from maintainers to take care of it. Right?

ntadej commented:
Yes, I wonder if we could submodule boost directly? Would it be that larger?

kiibimees commented:
> Yes, I wonder if we could submodule boost directly? Would it be that larger?

I thought about that. The trouble is that Boost repos are "source" repos but we need a "release" repo.

ntadej commented:
OK, then we need to fork it here, but it may take a bit to make that happen.

What is the timescale to get this into Qt 6.3? Are we already past the feature freeze? I wonder if we could use your fork for now...

kiibimees commented:
> What is the timescale to get this into Qt 6.3?

Unfortunately it doesn't got into Qt 6.3 (for political reasons), so I'm not in a super-hurry. Just that I would like to finish it and merge everything to Qt dev branch. At the moment there are a bunch on patches yet not merged in Qt Gerrit.

ntadej commented:
> > What is the timescale to get this into Qt 6.3?
> 
> Unfortunately it doesn't got into Qt 6.3 (for political reasons), so I'm not in a super-hurry. Just that I would like to finish it and merge everything to Qt dev branch. At the moment there are a bunch on patches yet not merged in Qt Gerrit.

That's unfortunate. Will it at least build against Qt 6.3 if I want to start porting my app to Qt 6 at some point?

To get this MR forward I propose the following (I can get it done this weekend):
 - As I just got rights, I create a personal repo for boost (which will be eventually moved under `maplibre`) and we use that.
 - I'll have another quick look of the changes, but I think we're good to go.

kiibimees commented:
> As I just got rights,
Very good. I can wait.

> Will it at least build against Qt 6.3 if I want to start porting my app to Qt 6 at some point?
Yes, the patches are actually based on Qt 6.3. It should compile as is. Just did it yesterday. For Qt 6.2 there is a need to revert one JSON/CBOR plugin change that was made after Qt 6.2 branch off.

ntadej commented:
> @ntadej , any news?

Sorry, was sick for the past two weeks. Let's wait until Wednesday when there is the steering committee meeting and we may already get a proper repo then.

ntadej commented:
While we're waiting for the repo: @kiibimees, did you try including QML Map using MapLibre in another QML element? I my Qt5 port I have issues that it is always rendered behind other QML elements. Also I have some issues with fonts.

kiibimees commented:
In Qt6 it works for me (screenshot added). Though, there is an issue with margin. Good that it came out now:
![image](https://user-images.githubusercontent.com/93144752/153197999-23264563-a71b-408d-8f8e-fe2581dcaad1.png)

ntadej commented:
@kiibimees, we now have https://github.com/maplibre/maplibre-gl-native-boost

Can you use that one? It uses @dpaulat minimal fix which is hopefully enough.

Also please rebase this MR so the Qt pipelines will run and I guess this one could be merged soon then.

kiibimees commented:
> Can you use that one?

Very good! But, perhaps this subrepo change could be a separate commit? And then I can rebase all my changes on top of that.

ntadej commented:
You mean a separate MR? Do you mind opening it, it will go in faster.

kiibimees commented:
> You mean a separate MR?

Sure, here it is: https://github.com/maplibre/maplibre-gl-native/pull/231

ntadej commented:
Thanks, merged!

Now the conflicts need to be resolved (rebase would be ideal).

kiibimees commented:
Yes, I will do that in due course...

kiibimees commented:
I made the rebase now and closed all issues but strange that GitHub still shows: "1 change requested"

ntadej commented:
> I made the rebase now and closed all issues but strange that GitHub still shows: "1 change requested"

I think I just need to sign-off, will do it later today (if I can).

ntadej commented:
Unfortunately static builds fail for iOS...

rinigus commented:
Do we actually have any tests for Linux / Qt builds? Can't see any or did I just miss it?

ntadej commented:
If it works with `make` and `ninja`, we can say that XCode is not supported, I suppose. Feel free to adapt the CI test.

ntadej commented:
> Do we actually have any tests for Linux / Qt builds? Can't see any or did I just miss it?

Not yet, pending on my TODO list (but from my experience, if it builds on macOS, it builds on Linux).

kiibimees commented:
Yes, I missed that detail. I guess it's your call because I'm not so familar with the CI setup. Using XCode generator and compined build is a very delicate thing in Qt. One can drown into EFFECTIVE_PLATFORM_NAME and all this magic that even in Qt 6 doesn't work too correctly.

ntadej commented:
I do not consider Qt support in this repo complete, so I propose you comment out the `iOS_static` in the test matrix.

ntadej commented:
Node CI is failing due to broken mirror (and probably needs the CI update to avoid this in the future). I see no way how node could be affected by the changes though so from my side it is safe to be merged.

I'm happy to get this merged as the Qt support is anyways still a bit rough and this will allow the work to continue faster (not to make the MR even longer than needed).

Unless there is any pressing issue from @roblabs or @birkskyum, I will merge this tomorrow.

rinigus commented:
Thanks for merging it! Are there any plans to start tagging Qt releases? Not sure why there are separate releases still for iOS and Android without making just a regular release for all platforms.

ntadej commented:
Once we have pipelines fully running for all platforms that we want to support, I do not see a reason why we could not tag Qt releases.

I think the reason for separate tags for iOS and Android is because the core functionality is not changing that much, but work is ongoing on the SDK side. There's also a question if you have a bugifix for Android, to you tag all other platforms as well?

rinigus commented:
You tag master branch, so you already tag it for all platforms. But I would tag bugfixes as just a regular release - not having that "Android" or "iOS" prefix. And as we do for other software, you just tag whole lib even if the bugfix is for some of the supported platforms.

ntadej commented:
Indeed I see no reason why we could not unify releases, but the repo needs to be cleaned-up a bit first.

kiibimees commented:
> This is now merged. Thanks again @kiibimees for all the work getting this updated for Qt 6!

Thank you!

> Note that I still plan to include the QML plugin in this repo (at least until Qt 6 is stable enough).

Yes, sure. It takes a little time until it's ready in Qt itself.

> Indeed I see no reason why we could not unify releases, but the repo needs to be cleaned-up a bit first.

I tend to agree. It has been a bit confusing for me as well that we have reparate releases to separate platforms.

kiibimees commented:
Thank you all, gentelmen!

Qt plugin (still work in progress though) can now be compiled and used from here:
https://codereview.qt-project.org/c/qt/qtlocation/+/382087
